### PR TITLE
Return cartesian product on multi match clauses

### PR DESCRIPTION
**Pull request by Josipmrden**
In multi match clauses, the result of 2 matches equals a cartesian product of rows. If both match clauses have filters, then there is a quadratic amount of filtering operations in the system.

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Provide the full content or a guide for the final git message

### CODE REVIEW

File: src/query/plan/pretty_print.cpp [hunk 1]
```diff
@@ -188,12 +187,28 @@ bool PlanPrinter::PreVisit(query::plan::Filter &op) {
...
 }
 
+bool PlanPrinter::PreVisit(query::plan::EdgeUniquenessFilter &op) {
+  WithPrintLn([&](auto &out) {
+    out << "* EdgeUniquenessFilter [";
+    utils::PrintIterable(out, op.previous_symbols_, ", ", [](auto &out, const auto &sym) { out << sym.name(); });
+    out << " != " << op.expand_symbol_.name() << "]";
+  });
+  return true;
+}
```
Comment (antepusic) [inline src/query/plan/pretty_print.cpp:197]: I suggest implementing `PlanPrinter::Previsit` for this operator with a `ToString` method, consistently with other operators that have detailed info (see [PR](https://github.com/memgraph/memgraph/pull/1204)).

File: tests/unit/query_plan.cpp [hunk 2]
```diff
@@ -280,8 +280,43 @@ TYPED_TEST(TestPlanner, MatchMultiPattern) {
...
+                                                    new ExpectExpand()};
+
+  auto symbol_table = memgraph::query::MakeSymbolTable(query);
+  auto planner = MakePlanner<TypeParam>(&dba, this->storage, symbol_table, query);
+  CheckPlan(planner.plan(), symbol_table, ExpectIndexedJoin(left_indexed_join_ops, right_indexed_join_ops),
+            ExpectEdgeUniquenessFilter(), ExpectProduce());
+
+  DeleteListContent(&left_indexed_join_ops);
+  DeleteListContent(&right_indexed_join_ops);
 }
```
Comment (antepusic) [inline tests/unit/query_plan.cpp:320]: What’s the reason as to why only one branch of `IndexedJoin` contains `ScanAllByLabelPropertyValue`, whereas the other one does `ScanAllByLabel`?

File: src/query/plan/preprocess.cpp [hunk 3]
```diff
@@ -55,35 +56,127 @@ void ForEachPattern(Pattern &pattern, std::function<void(NodeAtom *)> base,
...
+    matching.number_of_expansion_groups = matching.number_of_expansion_groups < expansion.expansion_group_id.AsUint()
+                                              ? expansion.expansion_group_id.AsUint()
+                                              : matching.number_of_expansion_groups;
+    next_expansion_group_id = ExpansionGroupId::FromUint(matching.number_of_expansion_groups + 1);
+  }
+
+  // By the time we finished assigning expansions, no expansion should have its expansion group ID unassigned
+  for (const auto &expansion : matching.expansions) {
+    MG_ASSERT(expansion.expansion_group_id.AsInt() != -1,
+              "Expansion expansion group ID is not assigned to the pattern!");
```
Comment (antepusic) [inline src/query/plan/preprocess.cpp:129]:
```suggestion
              "Expansion group ID is not assigned to the pattern!");
```

**Review by antepusic — CHANGES_REQUESTED** [review]
Good job! I’ve left a few comments below:

**Review by antoniofilipovic — CHANGES_REQUESTED** [review]
I checked PR and don't have many comments. Good work, left few suggestions otherwise it seems fine to me

**Review by antoniofilipovic — COMMENTED** [review]
I checked PR and don't have many comments. Good work, left few suggestions otherwise it seems fine to me

**Review by antepusic — CHANGES_REQUESTED** [review]
Having investigated, the checks are failing because `cartesian_product_enabled_` is turned off during unit test execution.

**Review by antepusic — COMMENTED** [review]
This looks good and very close to mergeable, and I’ve resolved some of my earlier comments. Before I hit approve, please respond to the remaining two comments I had left :smile:

**Review by antepusic — APPROVED** [review]
Everything looks good to me now, and the checks are now passing :rocket:

**Review by antoniofilipovic — APPROVED** [review]
Looks good to me

**Pull request merged**